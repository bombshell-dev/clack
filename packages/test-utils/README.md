# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
