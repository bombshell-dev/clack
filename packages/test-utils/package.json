{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
