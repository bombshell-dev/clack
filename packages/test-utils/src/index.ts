export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
