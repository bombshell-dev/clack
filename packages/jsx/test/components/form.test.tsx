import { MockReadable, MockWritable, nextTick } from '@clack/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';
import { Field, Form, Text } from '../../src/index.js';

describe('Form', () => {
	let input: MockReadable;
	let output: MockWritable;

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	test('renders and resolves object', async () => {
		const element = (
			<Form>
				<Field name="foo">
					<Text message="enter some text" />
				</Field>
			</Form>
		);
		const task = element.render({ input, output });

		input.emit('keypress', 'a', { name: 'a' });
		input.emit('keypress', 'b', { name: 'b' });
		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.deep.equal({
			foo: 'ab',
		});
		expect(output.buffer).toMatchSnapshot();
	});
	test('renders and resolves multiple fields', async () => {
		const element = (
			<Form>
				<Field name="foo">
					<Text message="enter some text" />
				</Field>
				<Field name="bar">
					<Text message="enter some other text" />
				</Field>
			</Form>
		);
		const task = element.render({ input, output });

		input.emit('keypress', 'a', { name: 'a' });
		input.emit('keypress', '', { name: 'return' });
		await nextTick();
		input.emit('keypress', 'b', { name: 'b' });
		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.deep.equal({
			foo: 'a',
			bar: 'b',
		});
		expect(output.buffer).toMatchSnapshot();
	});
});
