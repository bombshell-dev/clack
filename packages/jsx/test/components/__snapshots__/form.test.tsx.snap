// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form > renders and resolves multiple fields 1`] = `
[
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  enter some text
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  aâ–ˆ",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  enter some text
[90mâ”‚[39m  [2ma[22m",
  "
",
  "<cursor.show>",
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  enter some other text
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  bâ–ˆ",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  enter some other text
[90mâ”‚[39m  [2mb[22m",
  "
",
  "<cursor.show>",
]
`;

exports[`Form > renders and resolves object 1`] = `
[
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  enter some text
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  aâ–ˆ",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  abâ–ˆ",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  enter some text
[90mâ”‚[39m  [2mab[22m",
  "
",
  "<cursor.show>",
]
`;
